### DISCUSSION TITLE
The Art of the Minimal Reproducible Example

### DISCUSSION THREAD

**Discussion by TomJGooding** [Textualize/textual#3361] [General]
I recently discovered that Textualize founder @willmcgugan also started [label.dev](https://label.dev), which includes an article about [The Art of the Minimal Reproducible Example](https://label.dev/articles/minimal-reproducible-example/) which I thought worth sharing here.

When asking for help with a problem, you can greatly improve the chances of getting a quick answer by including a minimal reproducible example (MRE): a ***short*** piece of code that can reproduce the issue. It should be complete and self-contained, so anyone can just copy, paste and run the code.

The article also links to [Craft Minimal Bug Reports](https://matthewrocklin.com/minimal-bug-reports.html), which has more detailed advice on creating a MRE. Some highlights:

- Don't just post the file that you’re working with
- Is every line in your example absolutely necessary to reproduce the error?
- Provide complete tracebacks
- [Use Syntax Highlighting](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks)
- Test the code you’re about to provide to make sure it reproduces the problem

Reducing your code to a MRE will take some time and effort. but it "is good practice for solving bugs in general". 

> You may find after doing the work that you understand the issue better. In fact, it may reveal that the bug is not where you thought it was!"

Not every single problem needs a MRE and sometimes it just might not be possible. But if you do make the effort, an example will make people much more likely to help!

**Comment by davep** [comment]
Thanks Tom, a well-crafted writeup of the importance of MREs. Deserves a pin to the top of discussions.
